### PR TITLE
feat: add optional integrity verification for model artifacts

### LINKED ISSUE #761: [FEAT] Optionally Support integrity in model and wasm config
It is helpful tp optionally support integrity field in model and wasm config so we can do a bit of extra validation when fetching if users decide to setup the config to do so

### PR DESCRIPTION
**Pull request by hamzaydia**
## Summary

Adds optional [SRI (Subresource Integrity)](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) hash verification for model artifacts, as requested in #761.

When the `integrity` field is set on a `ModelRecord`, WebLLM verifies downloaded config, WASM, and tokenizer files against the provided SHA-256/384/512 hashes before loading. If a hash does not match, an `IntegrityError` is thrown (or a warning is logged when `onFailure: "warn"`).

### Usage

```typescript
const appConfig = {
  model_list: [{
    model: "https://huggingface.co/mlc-ai/Llama-3.2-1B-Instruct-q4f16_1-MLC",
    model_id: "Llama-3.2-1B-Instruct-q4f16_1-MLC",
    model_lib: "https://raw.githubusercontent.com/user/model-libs/main/model.wasm",
    integrity: {
      config: "sha256-<base64-hash>",
      model_lib: "sha256-<base64-hash>",
      tokenizer: { "tokenizer.json": "sha256-<base64-hash>" },
      onFailure: "error",
    },
  }],
};
```

### Design decisions

- **Zero new runtime dependencies** — uses only the Web Crypto API (`crypto.subtle.digest`) and the existing `loglevel` logger
- **Fully backwards compatible** — `integrity` is optional on `ModelRecord`, all fields within `ModelIntegrity` are optional, and omitting it entirely preserves existing behavior
- **Minimal diff to the loading pipeline** — config verification requires fetching as `arraybuffer` instead of `json` (only when integrity is set), WASM verification slots in after `fetchWasmSource()`, and tokenizer verification is added via a new optional parameter on `asyncLoadTokenizer()`
- **No changes to `prebuiltAppConfig`** — users generate SRI hashes for their own model deployments
- **No weight shard verification** — this would require changes to `@mlc-ai/web-runtime`'s `fetchTensorCache()`. For full model weight verification with resumable downloads and chunked verification, [`@verifyfetch/webllm`](https://github.com/hamzaydia/verifyfetch/tree/main/packages/webllm) can be used as a drop-in complement

### Changes

| File | Description |
|------|-------------|
| `src/integrity.ts` | **New.** Core module: `verifyIntegrity()`, `isValidSRI()`, `ModelIntegrity` interface, `SRIString`/`FileIntegrityMap` types |
| `src/error.ts` | Add `IntegrityError` class (follows existing error class pattern) |
| `src/config.ts` | Add optional `integrity?: ModelIntegrity` field to `ModelRecord` |
| `src/engine.ts` | Verify config + WASM in `reloadInternal()`, pass integrity to tokenizer loader |
| `src/cache_util.ts` | Verify tokenizer files in `asyncLoadTokenizer()` |
| `src/index.ts` | Export new types and utilities |
| `tests/integrity.test.ts` | **New.** 25 unit tests covering all hash algorithms, error paths, edge cases, and known reference hashes |
| `tests/cache_util.test.ts` | 8 new integration tests verifying the integrity parameter is correctly wired into `asyncLoadTokenizer` |
| `examples/integrity-verification/` | **New.** Working example demonstrating native integrity usage |
| `README.md` | New "Integrity Verification" section with usage example and SRI hash generation command |

### Test plan

- [x] All 25 unit tests pass (SHA-256/384/512, known hashes, error paths, edge cases)
- [x] All 8 integration tests pass (tokenizer verification wiring, error propagation)
- [x] Full test suite passes (179 tests, 16 suites, 0 failures)
- [x] `npx rollup -c` build succeeds
- [x] `npx eslint` and `npx prettier --check` pass on all modified files
- [x] Existing tests unaffected (backwards compatibility confirmed)

Closes #761

### CODE REVIEW

File: src/cache_util.ts [hunk 1]
```diff
@@ -132,6 +135,14 @@ export async function asyncLoadTokenizer(
...
     const url = new URL("tokenizer.json", baseUrl).href;
     const model = await modelCache.fetchWithCache(url, "arraybuffer");
+    if (integrity?.tokenizer?.["tokenizer.json"]) {
+      await verifyIntegrity(
+        model,
+        integrity.tokenizer["tokenizer.json"],
+        url,
+        integrity.onFailure,
+      );
+    }
```
[the remark this answers is not in the document]
Comment (hamzaydia) [reply src/cache_util.ts]: Done — extracted `verifyTokenizerIfNeeded()` helper that both code paths now call in e47276d.

File: src/engine.ts [hunk 2]
```diff
@@ -269,15 +270,36 @@ export class MLCEngine implements MLCEngineInterface {
...
+      curModelConfig = {
+        ...(await configCache.fetchWithCache(
+          configUrl,
+          "json",
+          this.reloadController?.signal,
+        )),
+        ...modelRecord.overrides,
+        ...chatOpts,
+      } as ChatConfig;
+    }
```
[the remark this answers is not in the document]
Comment (hamzaydia) [reply src/engine.ts]: Done — separated fetch/verify from the override merging. Both branches now set `configBase`, and the `...overrides, ...chatOpts` spread happens once at the end in e47276d.

File: src/engine.ts [hunk 3]
```diff
@@ -269,12 +270,30 @@ export class MLCEngine implements MLCEngineInterface {
...
+      );
+      configBase = JSON.parse(new TextDecoder().decode(configData));
+    } else {
+      configBase = (await configCache.fetchWithCache(
         configUrl,
         "json",
         this.reloadController?.signal,
-      )),
+      )) as Record<string, unknown>;
+    }
```
[the remark this answers is not in the document]
Comment (hamzaydia) [reply src/engine.ts]: Addressed in 8a20086 — config is now always fetched as arraybuffer and conditionally verified, eliminating the if/else branching.

File: README.md [hunk 4]
```diff
@@ -282,41 +282,106 @@ Then in the main logic, we register the service worker and create the engine usi
...
+You can generate SRI hashes for model files with:
+
+```bash
+# SHA-256
+openssl dgst -sha256 -binary <file> | openssl base64 -A | sed 's/^/sha256-/'
+# SHA-384
+openssl dgst -sha384 -binary <file> | openssl base64 -A | sed 's/^/sha384-/'
+# SHA-512
+openssl dgst -sha512 -binary <file> | openssl base64 -A | sed 's/^/sha512-/'
+```
```
[the remark this answers is not in the document]
Comment (hamzaydia) [reply README.md]: Addressed in 8a20086 — added a Windows compatibility note recommending the cross-platform npx verifyfetch sign command, with Git Bash and WSL as alternatives.

File: examples/integrity-verification/src/integrity_verification.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      },
+    ],
+  };
+
+  try {
+    const engine = await webllm.CreateMLCEngine(
+      "Llama-3.2-1B-Instruct-q4f16_1-MLC",
+      {
+        appConfig,
+        initProgressCallback: (report) => {
```
Comment (akaashrp) [inline examples/integrity-verification/src/integrity_verification.ts]: Type report with `webllm.InitProgressReport` (see `get_started.ts` for reference)

File: src/error.ts [hunk 6]
```diff
@@ -601,3 +601,19 @@ export class ReloadModelIdNotUniqueError extends Error {
     this.name = "ReloadModelIdNotUniqueError";
   }
 }
+
+export class IntegrityError extends Error {
+  constructor(
+    public readonly url: string,
```
Comment (akaashrp) [inline src/error.ts]: `public` is the default, so let's remove this

File: examples/integrity-verification/src/integrity_verification.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      );
+    } else {
+      setStatus("Error: " + error);
+    }
+  }
+}
+
+main();
+
+// ---------------------------------------------------------------------------
```
Comment (akaashrp) [inline examples/integrity-verification/src/integrity_verification.ts]: I’d prefer not to include this block in the example source.
1. It can be interpreted as an endorsement of a third-party package from WebLLM.
2. Our examples usually stay focused on native WebLLM usage and avoid extra promotional/contextual commentary.

I’m fine with mentioning this in the PR description with wording that clearly states this is informational.

File: tests/integrity.test.ts [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+      actual1 = (err as IntegrityError).actual;
+    }
+
+    try {
+      await verifyIntegrity(data2, wrongSRI, "https://example.com/2", "error");
+    } catch (err) {
+      actual2 = (err as IntegrityError).actual;
+    }
+
+    // The actual hashes should be different because the data is different
```
Comment (akaashrp) [inline tests/integrity.test.ts]: This comment and the one on line 302 are not necessary

File: tests/integrity.test.ts [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    }
+    const sri = `sha256-${btoa(binary)}`;
+
+    await expect(
+      verifyIntegrity(largeData.buffer, sri, "https://example.com/large.bin"),
+    ).resolves.toBeUndefined();
+  });
+
+  test("rejects when data is modified after hash computation", async () => {
+    // Compute hash of original data, then modify data and verify it fails
```
[the remark this answers is not in the document]
Comment (akaashrp) [reply tests/integrity.test.ts]: (including the ones on lines 266 and 267)

File: examples/integrity-verification/package.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  "devDependencies": {
+    "buffer": "^5.7.1",
+    "parcel": "^2.8.3",
+    "process": "^0.11.10",
+    "tslib": "^2.3.1",
+    "typescript": "^4.9.5",
+    "url": "^0.11.3"
+  },
+  "dependencies": {
+    "@mlc-ai/web-llm": "^0.2.80"
```
Comment (akaashrp) [inline examples/integrity-verification/package.json]: Change to 0.2.81
